snapshotter/benchmark_test.go

```diff
@@ -0,0 +1,206 @@
+	flag.StringVar(&dmRootPath, "dm.rootPath", "", "Root dir for devmapper snapshotter")
+	flag.StringVar(&overlayRootPath, "overlay.rootPath", "", "Root dir for overlay snapshotter")
+	flag.StringVar(&naiveRootPath, "naive.rootPath", "", "Root dir for naive snapshotter")
+}
+
+func BenchmarkNaive(b *testing.B) {
+	if naiveRootPath == "" {
+		b.Skip("naive snapshotter root dir must be provided")
+	}
+
+	snapshotter, err := naive.NewSnapshotter(context.Background(), naiveRootPath)
+	require.NoErrorf(b, err, "failed to create naive snapshotter")
+
+	defer func() {
+		err = snapshotter.Close()
+		assert.NoError(b, err)
+
+		err = os.RemoveAll(naiveRootPath)
+		assert.NoError(b, err)
+	}()
+
+	benchmarkSnapshotter(b, snapshotter)
+}
+
+func BenchmarkOverlay(b *testing.B) {
+	if overlayRootPath == "" {
+		b.Skip("overlay root dir must be provided")
+	}
+
+	snapshotter, err := overlay.NewSnapshotter(overlayRootPath)
+	require.NoErrorf(b, err, "failed to create overlay snapshotter")
+
+	defer func() {
+		err = snapshotter.Close()
+		assert.NoError(b, err)
+
+		err = os.RemoveAll(overlayRootPath)
+		assert.NoError(b, err)
+	}()
+
+	benchmarkSnapshotter(b, snapshotter)
+}
+
+func BenchmarkDeviceMapper(b *testing.B) {
+	if dmPoolDev == "" {
+		b.Skip("devmapper benchmark requires thin-pool device to be prepared in advance and provided")
+	}
+
+	if dmRootPath == "" {
+		b.Skip("devmapper snapshotter root dir must be provided")
+	}
+
+	config := &devmapper.Config{
+		PoolName:      dmPoolDev,
+		RootPath:      dmRootPath,
+		BaseImageSize: "16Mb",
+	}
+
+	ctx := context.Background()
+
+	snapshotter, err := devmapper.NewSnapshotter(ctx, config)
+	require.NoError(b, err)
+
+	defer func() {
+		err := snapshotter.ResetPool(ctx)
+		assert.NoError(b, err)
+
+		err = snapshotter.Close()
+		assert.NoError(b, err)
+
+		err = os.RemoveAll(dmRootPath)
+		assert.NoError(b, err)
+	}()
+
+	benchmarkSnapshotter(b, snapshotter)
+}
+
+// benchmarkSnapshotter tests snapshotter performance.
+// It writes 16 layers with randomly created, modified, or removed files.
+// Depending on layer index different sets of files are modified.
+func benchmarkSnapshotter(b *testing.B, snapshotter snapshots.Snapshotter) {
+	const (
+		layerCount    = 16
+		fileSizeBytes = int64(1 * 1024 * 1024) // 1 MB
+	)
+
+	var (
+		total      = 0
+		layers     = make([]fstest.Applier, layerCount)
+		layerIndex = int64(0)
+	)
+
+	for i := 0; i < layerCount; i++ {
+		appliers := makeApplier(i, fileSizeBytes)
+		layers[i] = fstest.Apply(appliers...)
+		total += len(appliers)
+	}
+
+	b.SetBytes(int64(total) * fileSizeBytes)
+	b.ResetTimer()
+
+	b.RunParallel(func(pb *testing.PB) {
+		ctx := context.Background()
+
+		for pb.Next() {
+			parent := ""
+
+			for l := 0; l < layerCount; l++ {
+				current := fmt.Sprintf("prepare-layer-%d", atomic.AddInt64(&layerIndex, 1))
+
+				mounts, err := snapshotter.Prepare(ctx, current, parent)
```

Comment:
Can you make sub-benchmarks for each of these steps?  The documentation for `b.RunParallel` says not to use `b.Run` inside (and there's no `pb.Run` as far as I can tell), but I'd like to get separate information for:

* how long it takes to prepare a snapshot (device activation?)
* how long it takes to write into the snapshot
* how long it takes to commit the snapshot